reactor-core/src/main/java/reactor/core/publisher/FluxIterable.java

```diff
@@ -119,18 +117,18 @@ static <T> void subscribe(CoreSubscriber<? super T> s, Iterator<? extends T> it,
 	 * is cancelled). Null to ignore.
 	 */
 	@SuppressWarnings("unchecked")
-	static <T> void subscribe(CoreSubscriber<? super T> s, Iterator<? extends T> it,
+	static <T> void subscribe(CoreSubscriber<? super T> s, Spliterator<? extends T> it,
 			boolean knownToBeFinite, @Nullable Runnable onClose) {
 		//noinspection ConstantConditions
 		if (it == null) {
 			Operators.error(s, new NullPointerException("The iterator is null"));
 			return;
 		}
 
-		boolean b;
+		long size;
 
 		try {
-			b = it.hasNext();
+			size = knownToBeFinite ? it.estimateSize() : -1;
```

Comment:
I'd love to see a comment explaining why this is feasible, e.g. "If the Spliterator is finite and we know it has not been traversed, the estimate is accurate." - but this raises a question - should we check whether the Spliterator has not been traversed yet? Is there a way to do so?

Reply:
I think it is irrelevant. At this stage, we just retrieved a new spliterator instance. So does it make sense to care about "check whether the Spliterator has not been traversed yet"? 

Also, the code it selves is taken from the method `Spliterator#getExactSizeIfKnown()` so this avoids extra work since we `knowToBeFinite`

Reply:
Right, I failed to record the fact the spliterator was retrieved in the constructor. Makes me think though: what happens in case of -1?

Reply:
it is not used explicitly. I tend to think that indeed we need to replace that with simplified code. e.g. 

```java
hasValues = !knownToBeFinite || it.estimateSize() != 0
```

Reply:
using the above we can say: ok if it is unknown to be finite - we can do nothing, assume has values, if it is known to be finite and estimated size is not 0 then it has values

Reply:
Sounds good, thanks.